### PR TITLE
JS: Refactor link & badge generation, use URLs (not string) for base URLs

### PR DESCRIPTION
**Pull request by yuvipanda**
This is two changes that were easier to make together

- BASE_URL and BADGE_BASE_URL are now URL objects rather than strings that are manipulated. With this done, we no longer use string manipulation for URLs anywhere!
- Both BASE_URL and BADGE_BASE_URL are now always set, as we had a bunch of code that was using BADGE_BASE_URL if available but falls back to BASE_URL + origin if it was not set.  This fallback
      is now implemented globally, and correctly.
- BASE_URL is also now always fully qualified, and we document that
      the python code ensures it has a trailing slash always.
- The function to make links and generate badge markup is moved into `@jupyterhub/binderhub-client` as it is reasonably generic and not super specific to our frontend alone. This also involves them not reading BASE_URL and BADGE_BASE_URL globally, but having that information be passed in. Tests are also added here to catch any future issues that may arise.
- Note for future fix - BADGE_BASE_URL is really PUBLIC_BASE_URL or similar, as it is used both for the location of the badge image (original intent) but also for the links we generate to share. This is relevant only for federation, where we want shared links to point to mybinder.org even though the API call itself may go to a specific member of the federation. I will do this deprecation + rename in a future PR so as to not make this PR bigger.

Here is a screen recording of me testing this out. You can see url generation + badge generation work fine for a couple cases, as well as building + launching.

https://github.com/jupyterhub/binderhub/assets/30430/f9cb387c-9308-4330-86e0-9f4d82bdfd7e

Ref https://github.com/jupyterhub/binderhub/issues/774

### CODE REVIEW

**Comment by yuvipanda** [comment]
Test failures still unrelated

File: binderhub/static/js/src/constants.js [hunk 1]
```diff
@@ -1,13 +1,22 @@
...
+ * @type {URL}
+ * Base URL of this binderhub installation.
+ *
+ * Guaranteed to have a trailing slash by the binderhub python configuration.
  */
-export const BASE_URL = $("#base-url").data().url;
+export const BASE_URL = new URL(
+  document.getElementById("base-url").dataset.url,
+  document.location.origin,
+);
```
[the remark this answers is not in the document]
Comment (yuvipanda) [reply binderhub/static/js/src/constants.js:10]: @consideRatio The binderhub python code guarantees that base_url will start with a `/` as well (https://github.com/jupyterhub/binderhub/blob/main/binderhub/app.py#L167), so `location.origin` and `location.href` are equivalent here (that's the behavior of the URL constructor). I've added a comment noting that a leading and trailing slash is enforced.

The way you would host something under a prefix *is* to set `base_url` to the prefix, so this is the functionality that allows hosting under prefixes!

[the remark this answers is not in the document]
Comment (consideRatio) [reply binderhub/static/js/src/constants.js:10]: Thanks for clarifying this @yuvipanda!

**Review by consideRatio — APPROVED** [review]

**Comment by consideRatio** [comment]
Like in https://github.com/jupyterhub/binderhub/pull/1773#issuecomment-1759728662

> I think its a good call to opt for a merge at this point even though I'm not so confident reviewing this, it helps with Yuvi's contribution momentum and we aren't introducing new API or similar that we commit to long term in this PR either.

I've done my best to review this, and I think it looks good!

**Comment by yuvipanda** [comment]
Thanks @consideRatio, I appreciate it! When both these PRs are merged, I'll roll them out to mybinder.org and keep an eye out.

**Pull request merged**